instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java

```diff
@@ -103,55 +106,116 @@
+     * Constructs a Searcher from an existing {@link Index}.
      *
      * @param index an Index initialized and eventually configured.
      * @return the new instance.
      */
     @SuppressWarnings({"WeakerAccess", "unused"}) // For library users
     public static Searcher create(@NonNull final Index index) {
+        final String key = index.getIndexName();
+        Searcher instance = instances.get(key);
         if (instance == null) {
-            instance = new Searcher(index);
+            instance = new Searcher(index, key);
+            instances.put(key, instance);
+        } else {
+            throw new IllegalStateException("There is already a Searcher for index " + key + ", you must specify a variant.");
         }
         return instance;
     }
 
     /**
-     * Constructs an helper, creating its {@link Searcher#index} and {@link Searcher#client} with the given parameters.
+     * Constructs the Searcher from an existing {@link Index}, eventually replacing the existing searcher for {@code variant}.
      *
-     * @param appId     Your Algolia Application ID.
-     * @param apiKey    A search-only API Key. (never use API keys that could modify your records! see https://www.algolia.com/doc/guides/security/api-keys)
-     * @param indexName An index to target.
+     * @param variant an identifier to differentiate this Searcher from eventual others using the same index. Defaults to the index's name.
+     * @param index   an Index initialized and eventually configured.
      * @return the new instance.
      */
     @SuppressWarnings({"WeakerAccess", "unused"}) // For library users
-    public static Searcher create(@NonNull final String appId, @NonNull final String apiKey, @NonNull final String indexName) {
-        return create(new Client(appId, apiKey).getIndex(indexName));
+    public static Searcher create(@NonNull final Index index, @NonNull String variant) {
+        boolean shouldReplace = false;
+        Searcher instance = instances.get(variant);
+        if (instance != null) {
+            shouldReplace = instance.getIndex() != index;
+        }
+        if (instance == null || shouldReplace) {
+            instance = new Searcher(index, variant);
+            instances.put(variant, instance);
+        }
+        return instance;
     }
 
-    private Searcher(@NonNull final Index index) {
+    private Searcher(@NonNull final Index index, @NonNull String variant) {
         this.index = index;
         this.client = index.getClient();
+        this.variant = variant;
         query = new Query();
         final LibraryVersion version = new LibraryVersion("InstantSearch Android", String.valueOf(BuildConfig.VERSION_NAME));
         if (!Arrays.asList(client.getUserAgents()).contains(version)) {
             client.addUserAgent(version);
         }
     }
 
+    /**
+     * Destroys all Searcher instances.
+     */
+    public static void destroyAll() {
```

Comment:
nit: suggestion: `clearAll`

Reply:
Open for debate: I called it `destroyAll` as it's meant to be used in the `Activity`/`Fragment`'s **`onDestroy`** method.

Reply:
@PLNech @spinach destroy is the Android specific name. We should go with that